runbot_gitlab/models/runbot_repo.py

```diff
@@ -0,0 +1,335 @@
+
+        branches = {}
+        if not self.mr_only:
+            for branch in self._query_gitlab_api(
+                    'projects/%s/repository/branches' % project['id']):
+                branches[branch['commit']['id']] = branch
+        for merge_request in self._query_gitlab_api(
+                'projects/%s/merge_requests' % project['id'],
+                get_data={'state': 'opened'}):
+            # strangely enough, we can't query the commits with
+            # projects/%s/merge_request/%s/commits, so we fetch the
+            # source branch and get its last commit
+            branch = self._query_gitlab_api(
+                'projects/%s/repository/branches/%s' % (
+                    merge_request['source_project_id'],
+                    merge_request['source_branch']))
+            if not branch:
+                continue
+            branch['merge_request'] = merge_request
+            branches[branch['commit']['id']] = branch
+
+        builds_created = self.env['runbot.build'].browse([])
+
+        # create or adapt builds
+        for sha, branch in branches.iteritems():
+            commit = branch['commit']
+            date = dateutil.parser.parse(commit['committed_date'])
+            try:
+                author = commit['author']['name']
+            except KeyError:
+                author = commit.get('author_name')
+            try:
+                committer = commit['committer']['name']
+            except KeyError:
+                committer = commit.get('committer_name')
+
+            branch_name = 'refs/heads/%s' % branch['name']
+
+            # Create or get branch
+            branch_ids = self.env['runbot.branch'].search([
+                ('repo_id', '=', self.id),
+                ('name', '=', branch_name),
+            ])
+            if not branch_ids:
+                logger.debug('repo %s found new merge request or build %s',
+                             self.name, sha)
+                branch_ids = self.env['runbot.branch'].create({
+                    'repo_id': self.id,
+                    'name': branch_name,
+                    'project_id': project['id'],
+                    'merge_request_id':
+                    branch.get('merge_request', {}).get('iid'),
+                })
+            else:
+                # there's a chance this was a non-mr-branch before
+                branch_ids.filtered(lambda x: not x.merge_request_id)\
+                    .write({
+                        'merge_request_id':
+                        branch.get('merge_request', {}).get('iid'),
+                    })
+                # in this case, this branch's builds were skipped, undo that
+                if self.mr_only:
+                    self.env['runbot.build'].search([
+                        ('branch_id', 'in', branch_ids.ids),
+                        ('result', '=', 'skipped'),
+                        ('state', '=', 'done'),
+                    ], limit=1).force()
```

Comment:
Probably I don't understand the whole updating logic, but isn't there a chance that this line would also force the build of no-MR branches?